cmd/pulumi-test-language/providers/config_grpc_provider.go

```diff
@@ -0,0 +1,331 @@
+			},
+		},
+		Types: types,
+		Resources: map[string]pschema.ResourceSpec{
+			fmt.Sprintf("%s:index:ConfigGetter", p.Pkg()): {
+				ObjectTypeSpec: pschema.ObjectTypeSpec{
+					Properties: map[string]pschema.PropertySpec{
+						"config": {
+							TypeSpec: pschema.TypeSpec{Type: "string"},
+						},
+					},
+					Required: []string{"config"},
+				},
+			},
+		},
+		Functions: map[string]pschema.FunctionSpec{},
+		Language: map[string]pschema.RawMessage{
+			"nodejs": []byte(`{"respectSchemaVersion": true}`),
+		},
+	}
+
+	toSecretSchema := p.generateSchema(types, 0, 0)
+	allProps := []string{}
+	for k := range toSecretSchema {
+		allProps = append(allProps, k)
+	}
+
+	schema.Functions[fmt.Sprintf("%s:index:toSecret", p.Pkg())] = pschema.FunctionSpec{
+		Inputs: &pschema.ObjectTypeSpec{
+			Properties: toSecretSchema,
+		},
+		Outputs: &pschema.ObjectTypeSpec{
+			Properties: toSecretSchema,
+			Required:   allProps,
+		},
+	}
+
+	return schema
+}
+
+func (p *ConfigGrpcProvider) NewProviderServer() pulumirpc.ResourceProviderServer {
+	return &configGrpcProviderServer{
+		ResourceProviderServer: plugin.NewProviderServer(p),
+	}
+}
+
+func (p *ConfigGrpcProvider) GetSchema(
+	context.Context, plugin.GetSchemaRequest,
+) (plugin.GetSchemaResponse, error) {
+	schema := p.schema()
+	schemaBytes, err := json.Marshal(schema)
+	if err != nil {
+		return plugin.GetSchemaResponse{}, err
+	}
+	return plugin.GetSchemaResponse{Schema: schemaBytes}, nil
+}
+
+func (p *ConfigGrpcProvider) GetPluginInfo(context.Context) (workspace.PluginInfo, error) {
+	ver := semver.MustParse(p.version())
+	return workspace.PluginInfo{
+		Version: &ver,
+	}, nil
+}
+
+func (p *ConfigGrpcProvider) Check(
+	_ context.Context, req plugin.CheckRequest,
+) (plugin.CheckResponse, error) {
+	return plugin.CheckResponse{Properties: req.News}, nil
+}
+
+func (p *ConfigGrpcProvider) CheckConfig(
+	_ context.Context, req plugin.CheckConfigRequest,
+) (plugin.CheckConfigResponse, error) {
+	return plugin.CheckConfigResponse{Properties: req.News}, nil
+}
+
+func (p *ConfigGrpcProvider) Configure(
+	context.Context, plugin.ConfigureRequest,
+) (plugin.ConfigureResponse, error) {
+	return plugin.ConfigureResponse{}, nil
+}
+
+func (p *ConfigGrpcProvider) DiffConfig(
+	context.Context, plugin.DiffConfigRequest,
+) (plugin.DiffConfigResponse, error) {
+	return plugin.DiffResult{}, nil
+}
+
+func (p *ConfigGrpcProvider) Invoke(
+	ctx context.Context, req plugin.InvokeRequest,
+) (plugin.InvokeResponse, error) {
+	switch {
+	case string(req.Tok) == fmt.Sprintf("%s:index:toSecret", p.Pkg()):
+		secreted := req.Args.Copy()
+		for k, v := range secreted {
+			secreted[k] = resource.MakeSecret(v)
+		}
+		return plugin.InvokeResponse{Properties: secreted}, nil
+	default:
+		return plugin.InvokeResponse{}, errors.New("Unknown function")
+	}
+}
+
+// This lower level implementation should be used at runtime specifically to test at the lower level, since not all
```

Comment:
I think I'd split the custom provider server implementation out into something like `providers/custom_logging_server.go` or the like, since I could see it being useful elsewhere. I'd probably rename it and the related property/methods too, since it's capturing more than just `Configure` calls at this point. Indeed, we don't have to (and probably shouldn't for the sake of a granular PR) do it now but I could see it being useful to extend it to capture all calls, really.